File: server/models/Category.js
const mongoose = require('mongoose');

const categorySchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: [true, 'Category name is required'],
      trim: true,
      unique: true,
      maxlength: [50, 'Category name cannot be more than 50 characters'],
    },
    slug: {
      type: String,
      required: true,
      unique: true,
      lowercase: true,
    },
    description: {
      type: String,
      maxlength: [500, 'Description cannot be more than 500 characters'],
    },
    image: { type: String, default: '' },
    icon: { type: String, default: '' },
    parent: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Category',
      default: null,
    },
    level: { type: Number, default: 0 },
    isActive: { type: Boolean, default: true },
    sortOrder: { type: Number, default: 0 },
    seoTitle: String,
    seoDescription: String,
    seoKeywords: [String],
  },
  { timestamps: true, toJSON: { virtuals: true }, toObject: { virtuals: true } }
);

// Virtual populate 'children' thay vì lưu array ObjectId
categorySchema.virtual('children', {
  ref: 'Category',
  localField: '_id',
  foreignField: 'parent',
  justOne: false,
});

// Auto-generate slug
categorySchema.pre('save', function (next) {
  if (this.isModified('name') && !this.slug) {
    this.slug = this.name
      .toLowerCase()
      .replace(/[^a-z0-9 -]/g, '')
      .replace(/\s+/g, '-')
      .replace(/-+/g, '-')
      .replace(/^-|-$/g, '');
  }
  next();
});

// Update level before save
categorySchema.pre('save', function (next) {
  this.level = this.parent ? 1 : 0;
  next();
});

// Auto populate children khi truy vấn
function autopopulate(next) {
  this.populate('children');
  next();
}
categorySchema.pre('find', autopopulate);
categorySchema.pre('findOne', autopopulate);

module.exports = mongoose.model('Category', categorySchema);

File: app/lib/models/category.dart
class Category {
  final String id;
  final String name;
  final String? slug;
  final String? image;
  final String? icon;
  final List<Category> children;

  Category({
    required this.id,
    required this.name,
    this.slug,
    this.image,
    this.icon,
    List<Category>? children,
  }) : children = children ?? [];

  factory Category.fromJson(Map<String, dynamic> json) {
    // Lấy _id từ MongoDB: json['_id']['\$oid'] hoặc string thẳng
    final rawId = json['_id'];
    String idValue;
    if (rawId is Map && rawId.containsKey(r'$oid')) {
      idValue = rawId[r'$oid'];
    } else if (rawId is String) {
      idValue = rawId;
    } else {
      idValue = '';
    }

    var childrenJson = json['children'] as List<dynamic>?;

    return Category(
      id: idValue,
      name: json['name'] ?? '',
      slug: json['slug'],
      image: json['image'],
      icon: json['icon'],
      children: childrenJson != null
          ? childrenJson.map((c) => Category.fromJson(c)).toList()
          : [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'slug': slug,
      'image': image,
      'icon': icon,
      'children': children.map((c) => c.toJson()).toList(),
    };
  }
}
